Sources/Logging/MetadataProvider.swift

```diff
@@ -0,0 +1,112 @@
+    /// to all logs emitted by a logger.
+    ///
+    /// ### Example
+    /// A metadata provider may be used to automatically inject metadata such as
+    /// trace IDs:
+    ///
+    /// ```swift
+    /// import Tracing // https://github.com/apple/swift-distributed-tracing
+    ///
+    /// let metadataProvider = MetadataProvider {
+    ///     guard let traceID = Baggage.current?.traceID else { return nil }
+    ///     return ["traceID": "\(traceID)"]
+    /// }
+    /// let logger = Logger(label: "example", metadataProvider: metadataProvider)
+    /// var baggage = Baggage.topLevel
+    /// baggage.traceID = 42
+    /// Baggage.withValue(baggage) {
+    ///     logger.info("hello") // automatically includes ["traceID": "42"] metadata
+    /// }
+    /// ```
+    public struct MetadataProvider: _SwiftLogSendable {
+        /// A default no-op metadata provider, which always returns empty metadata.
+        public static var noop: MetadataProvider {
+            return MetadataProvider { [:] }
+        }
+
+        /// Return the ``Logger/MetadataProvider-swift.struct`` that was configured during ``LoggingSystem/bootstrap(_:)-8ffrb``.
+        public static var bootstrapped: MetadataProvider {
+            return LoggingSystem.metadataProvider
+        }
+
+        #if swift(>=5.5) && canImport(_Concurrency)
+        public typealias Function = @Sendable() -> Metadata
+        #else
+        public typealias Function = () -> Metadata
+        #endif
+
+        /// Provide ``Logger.Metadata`` from current context.
+        @usableFromInline
+        internal let _provideMetadata: MetadataProvider.Function
+
+        /// Create a new `MetadataProvider`.
+        ///
+        /// - Parameter provideMetadata: A closure extracting metadata from a given `Baggage`.
+        public init(_ provideMetadata: @escaping MetadataProvider.Function) {
+            self._provideMetadata = provideMetadata
+        }
+
+        /// Invoke the metadata provider and return the generated contextual ``Logger/Metadata``.
+        public func provideMetadata() -> Metadata {
+            return self._provideMetadata()
+        }
+    }
+}
+
+extension Logger.MetadataProvider {
+    /// A pseudo-`MetadataProvider` that can be used to merge metadata from multiple other `MetadataProvider`s.
+    ///
+    /// ### Merging conflicting keys
+    ///
+    /// `MetadataProvider`s are invoked left to right in the order specified in the `providers` argument.
+    /// In case multiple providers try to add a value for the same key, the last provider "wins" and its value is being used.
+    ///
+    /// - Parameter providers: An array of `MetadataProvider`s to delegate to. The array must not be empty.
+    /// - Returns: A pseudo-`MetadataProvider` merging metadata from the given `MetadataProvider`s.
+    public static func multiplex(_ providers: [Logger.MetadataProvider]) -> Logger.MetadataProvider {
+        precondition(!providers.isEmpty, "providers MUST NOT be empty")
```

Comment:
Any reason to not just return `.noop` in this case?

Reply:
I replicated the behavior of `MultiplexLogHandler` with this. It should be an `assert` instead of a `precondition` though: https://github.com/apple/swift-log/blob/d86803715f1251129063c6a93f3f1ff4dfc1b2aa/Sources/Logging/Logging.swift#LL938C9-L938C15